<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · GBCore.jl</title><meta name="title" content="Home · GBCore.jl"/><meta property="og:title" content="Home · GBCore.jl"/><meta property="twitter:title" content="Home · GBCore.jl"/><meta name="description" content="Documentation for GBCore.jl."/><meta property="og:description" content="Documentation for GBCore.jl."/><meta property="twitter:description" content="Documentation for GBCore.jl."/><meta property="og:url" content="https://GenomicBreeding.github.io/GBCore.jl/"/><meta property="twitter:url" content="https://GenomicBreeding.github.io/GBCore.jl/"/><link rel="canonical" href="https://GenomicBreeding.github.io/GBCore.jl/"/><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="search_index.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>GBCore.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/GenomicBreeding/GBCore.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/GenomicBreeding/GBCore.jl/blob/main/docs/src/index.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="GBCore"><a class="docs-heading-anchor" href="#GBCore">GBCore</a><a id="GBCore-1"></a><a class="docs-heading-anchor-permalink" href="#GBCore" title="Permalink"></a></h1><p>Documentation for <a href="https://github.com/GenomicBreeding/GBCore.jl">GBCore</a>.</p><ul><li><a href="#GBCore.AbstractGB"><code>GBCore.AbstractGB</code></a></li><li><a href="#GBCore.CV"><code>GBCore.CV</code></a></li><li><a href="#GBCore.Fit"><code>GBCore.Fit</code></a></li><li><a href="#GBCore.Genomes"><code>GBCore.Genomes</code></a></li><li><a href="#GBCore.Phenomes"><code>GBCore.Phenomes</code></a></li><li><a href="#GBCore.SimulatedEffects"><code>GBCore.SimulatedEffects</code></a></li><li><a href="#GBCore.TEBV"><code>GBCore.TEBV</code></a></li><li><a href="#GBCore.Trials"><code>GBCore.Trials</code></a></li><li><a href="#Base.:==-Tuple{Genomes, Genomes}"><code>Base.:==</code></a></li><li><a href="#Base.:==-Tuple{Trials, Trials}"><code>Base.:==</code></a></li><li><a href="#Base.:==-Tuple{Fit, Fit}"><code>Base.:==</code></a></li><li><a href="#Base.:==-Tuple{TEBV, TEBV}"><code>Base.:==</code></a></li><li><a href="#Base.:==-Tuple{Phenomes, Phenomes}"><code>Base.:==</code></a></li><li><a href="#Base.:==-Tuple{CV, CV}"><code>Base.:==</code></a></li><li><a href="#Base.:==-Tuple{SimulatedEffects, SimulatedEffects}"><code>Base.:==</code></a></li><li><a href="#Base.filter-Tuple{Genomes, Float64}"><code>Base.filter</code></a></li><li><a href="#Base.filter-Tuple{Genomes}"><code>Base.filter</code></a></li><li><a href="#Base.filter-Tuple{Phenomes}"><code>Base.filter</code></a></li><li><a href="#Base.hash-Tuple{Trials, UInt64}"><code>Base.hash</code></a></li><li><a href="#Base.hash-Tuple{TEBV, UInt64}"><code>Base.hash</code></a></li><li><a href="#Base.hash-Tuple{Fit, UInt64}"><code>Base.hash</code></a></li><li><a href="#Base.hash-Tuple{Genomes, UInt64}"><code>Base.hash</code></a></li><li><a href="#Base.hash-Tuple{Phenomes, UInt64}"><code>Base.hash</code></a></li><li><a href="#Base.hash-Tuple{CV, UInt64}"><code>Base.hash</code></a></li><li><a href="#Base.hash-Tuple{SimulatedEffects, UInt64}"><code>Base.hash</code></a></li><li><a href="#Base.merge-Tuple{Phenomes, Phenomes}"><code>Base.merge</code></a></li><li><a href="#Base.merge-Tuple{Genomes, Genomes}"><code>Base.merge</code></a></li><li><a href="#Base.merge-Tuple{Genomes, Phenomes}"><code>Base.merge</code></a></li><li><a href="#Base.sum-Tuple{SimulatedEffects}"><code>Base.sum</code></a></li><li><a href="#GBCore.addcompositetrait-Tuple{Phenomes}"><code>GBCore.addcompositetrait</code></a></li><li><a href="#GBCore.addcompositetrait-Tuple{Trials}"><code>GBCore.addcompositetrait</code></a></li><li><a href="#GBCore.analyse"><code>GBCore.analyse</code></a></li><li><a href="#GBCore.analyse-Tuple{DataFrames.DataFrame}"><code>GBCore.analyse</code></a></li><li><a href="#GBCore.checkdims-Tuple{TEBV}"><code>GBCore.checkdims</code></a></li><li><a href="#GBCore.checkdims-Tuple{Phenomes}"><code>GBCore.checkdims</code></a></li><li><a href="#GBCore.checkdims-Tuple{Genomes}"><code>GBCore.checkdims</code></a></li><li><a href="#GBCore.checkdims-Tuple{SimulatedEffects}"><code>GBCore.checkdims</code></a></li><li><a href="#GBCore.checkdims-Tuple{Fit}"><code>GBCore.checkdims</code></a></li><li><a href="#GBCore.checkdims-Tuple{CV}"><code>GBCore.checkdims</code></a></li><li><a href="#GBCore.checkdims-Tuple{Trials}"><code>GBCore.checkdims</code></a></li><li><a href="#GBCore.clone-Tuple{TEBV}"><code>GBCore.clone</code></a></li><li><a href="#GBCore.clone-Tuple{Phenomes}"><code>GBCore.clone</code></a></li><li><a href="#GBCore.clone-Tuple{Genomes}"><code>GBCore.clone</code></a></li><li><a href="#GBCore.clone-Tuple{Trials}"><code>GBCore.clone</code></a></li><li><a href="#GBCore.clone-Tuple{Fit}"><code>GBCore.clone</code></a></li><li><a href="#GBCore.clone-Tuple{CV}"><code>GBCore.clone</code></a></li><li><a href="#GBCore.countlevels-Tuple{DataFrames.DataFrame}"><code>GBCore.countlevels</code></a></li><li><a href="#GBCore.dimensions-Tuple{TEBV}"><code>GBCore.dimensions</code></a></li><li><a href="#GBCore.dimensions-Tuple{Phenomes}"><code>GBCore.dimensions</code></a></li><li><a href="#GBCore.dimensions-Tuple{Genomes}"><code>GBCore.dimensions</code></a></li><li><a href="#GBCore.dimensions-Tuple{Trials}"><code>GBCore.dimensions</code></a></li><li><a href="#GBCore.distances-Tuple{Genomes}"><code>GBCore.distances</code></a></li><li><a href="#GBCore.distances-Tuple{Phenomes}"><code>GBCore.distances</code></a></li><li><a href="#GBCore.extractphenomes-Tuple{Trials}"><code>GBCore.extractphenomes</code></a></li><li><a href="#GBCore.extractphenomes-Tuple{TEBV}"><code>GBCore.extractphenomes</code></a></li><li><a href="#GBCore.loci-Tuple{Genomes}"><code>GBCore.loci</code></a></li><li><a href="#GBCore.loci_alleles-Tuple{Genomes}"><code>GBCore.loci_alleles</code></a></li><li><a href="#GBCore.plot-Tuple{TEBV}"><code>GBCore.plot</code></a></li><li><a href="#GBCore.plot"><code>GBCore.plot</code></a></li><li><a href="#GBCore.plot-Tuple{Trials}"><code>GBCore.plot</code></a></li><li><a href="#GBCore.plot-Tuple{Phenomes}"><code>GBCore.plot</code></a></li><li><a href="#GBCore.plot"><code>GBCore.plot</code></a></li><li><a href="#GBCore.simulateeffects-Tuple{}"><code>GBCore.simulateeffects</code></a></li><li><a href="#GBCore.simulategenomes-Tuple{}"><code>GBCore.simulategenomes</code></a></li><li><a href="#GBCore.simulategenomiceffects-Tuple{}"><code>GBCore.simulategenomiceffects</code></a></li><li><a href="#GBCore.simulatetrials-Tuple{}"><code>GBCore.simulatetrials</code></a></li><li><a href="#GBCore.slice-Tuple{Phenomes}"><code>GBCore.slice</code></a></li><li><a href="#GBCore.slice-Tuple{Trials}"><code>GBCore.slice</code></a></li><li><a href="#GBCore.slice-Tuple{Genomes}"><code>GBCore.slice</code></a></li><li><a href="#GBCore.summarise-Tuple{Vector{CV}}"><code>GBCore.summarise</code></a></li><li><a href="#GBCore.tabularise"><code>GBCore.tabularise</code></a></li><li><a href="#GBCore.tabularise-Tuple{Phenomes}"><code>GBCore.tabularise</code></a></li><li><a href="#GBCore.tabularise-Tuple{Trials}"><code>GBCore.tabularise</code></a></li><li><a href="#GBCore.tabularise-Tuple{Vector{CV}}"><code>GBCore.tabularise</code></a></li><li><a href="#GBCore.trialsmodelsfomulae!-Tuple{DataFrames.DataFrame}"><code>GBCore.trialsmodelsfomulae!</code></a></li><li><a href="#GBCore.@string2formula-Tuple{Any}"><code>GBCore.@string2formula</code></a></li><li><a href="#GBCore.@stringevaluation-Tuple{Any}"><code>GBCore.@stringevaluation</code></a></li></ul><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.AbstractGB" href="#GBCore.AbstractGB"><code>GBCore.AbstractGB</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Abstract super type for all GenomicBreeding.jl custom types</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/all_structs.jl#L1-L3">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.CV" href="#GBCore.CV"><code>GBCore.CV</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Cross-validation struct</strong></p><p>Contains genomic prediction cross-validation details.</p><p><strong>Fields</strong></p><ul><li><code>replication</code>: replication name</li><li><code>fold</code>: fold name</li><li><code>fit</code>: genomic prediction model fit on the training set</li><li><code>validation_populations</code>: vector of validation populations corresponding to each validation entry</li><li><code>validation_entries</code>: corresponding vector of entries in the validation population/s</li><li><code>validation_y_true</code>: corresponding vector of observed phenotypes in the validation population/s</li><li><code>validation_y_pred</code>: corresponding vector of predicted phenotypes in the validation population/s</li><li><code>metrics</code>: dictionary of genomic prediction accuracy metrics on the validation population/s</li></ul><p><strong>Constructor</strong></p><p>Uses the default contructor.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/all_structs.jl#L341-L358">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.Fit" href="#GBCore.Fit"><code>GBCore.Fit</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Genomic prediction model fit</strong></p><p>Contains genomic prediction model fit details.</p><p><strong>Fields</strong></p><ul><li><code>model</code>: name of the genomic prediction model used</li><li><code>b_hat_labels</code>: names of the loci-alleles used</li><li><code>b_hat</code>: effects of the loci-alleles</li><li><code>trait</code>: name of the trait</li><li><code>entries</code>: names of the entries used in the current cross-validation replication and fold</li><li><code>populations</code>: names of the populations used in the current cross-validation replication and fold</li><li><code>y_true</code>: corresponding observed phenotype values</li><li><code>y_pred</code>: corresponding predicted phenotype values</li><li><code>metrics</code>: dictionary of genomic prediction accuracy metrics, inluding Pearson&#39;s correlation, mean absolute error and root mean-squared error</li></ul><p><strong>Constructor</strong></p><pre><code class="language-julia hljs">Fit(; n=1, l=10)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/all_structs.jl#L294-L314">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.Genomes" href="#GBCore.Genomes"><code>GBCore.Genomes</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Genomes struct</strong></p><p>Containes unique entries and loci_alleles where allele frequencies can have missing values</p><p><strong>Constructor</strong></p><pre><code class="language-julia hljs">Genomes(; n::Int64 = 1, p::Int64 = 2)</code></pre><p><strong>Fields</strong></p><ul><li><code>entries</code>: names of the <code>n</code> entries or samples</li><li><code>populations</code>: name/s of the population/s each entries or samples belong to</li><li><code>loci_alleles</code>: names of the <code>p</code> loci-alleles combinations (<code>p</code> = <code>l</code> loci x <code>a-1</code> alleles) including the </li></ul><p>chromsome or scaffold name, position, all alleles, and current allele separated by tabs (&quot;\t&quot;)</p><ul><li><code>allele_frequencies</code>: <code>n x p</code> matrix of allele frequencies between 0 and 1 which can have missing values</li><li><code>mask</code>: <code>n x p</code> matrix of boolean mask for selective analyses and slicing</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = Genomes(n=2, p=2)
Genomes([&quot;&quot;, &quot;&quot;], [&quot;&quot;, &quot;&quot;], [&quot;&quot;, &quot;&quot;], Union{Missing, Float64}[missing missing; missing missing], Bool[1 1; 1 1])

julia&gt; fieldnames(Genomes)
(:entries, :populations, :loci_alleles, :allele_frequencies, :mask)

julia&gt; genomes.entries = [&quot;entry_1&quot;, &quot;entry_2&quot;];

julia&gt; genomes.populations = [&quot;pop_1&quot;, &quot;pop_1&quot;];

julia&gt; genomes.loci_alleles = [&quot;chr1\t12345\tA|T\tA&quot;, &quot;chr2\t678910\tC|D\tD&quot;];

julia&gt; genomes.allele_frequencies = [0.5 0.25; 0.9 missing];

julia&gt; genomes.mask = [true true; true false];

julia&gt; genomes
Genomes([&quot;entry_1&quot;, &quot;entry_2&quot;], [&quot;pop_1&quot;, &quot;pop_1&quot;], [&quot;chr1\t12345\tA|T\tA&quot;, &quot;chr2\t678910\tC|D\tD&quot;], Union{Missing, Float64}[0.5 0.25; 0.9 missing], Bool[1 1; 1 0])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/all_structs.jl#L7-L46">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.Phenomes" href="#GBCore.Phenomes"><code>GBCore.Phenomes</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Phenomes struct</strong></p><p>Constains unique entries and traits where phenotype data can have missing values</p><p><strong>Constructor</strong></p><pre><code class="language-julia hljs">Phenomes(; n::Int64 = 1, t::Int64 = 2)</code></pre><p><strong>Fields</strong></p><ul><li><code>entries</code>: names of the <code>n</code> entries or samples</li><li><code>populations</code>: name/s of the population/s each entries or samples belong to</li><li><code>traits</code>: names of the <code>t</code> traits</li><li><code>phenotypes</code>: <code>n x t</code> matrix of numeric (<code>R</code>) phenotype data which can have missing values</li><li><code>mask</code>: <code>n x t</code> matrix of boolean mask for selective analyses and slicing</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; phenomes = Phenomes(n=2, t=2)
Phenomes([&quot;&quot;, &quot;&quot;], [&quot;&quot;, &quot;&quot;], [&quot;&quot;, &quot;&quot;], Union{Missing, Float64}[missing missing; missing missing], Bool[1 1; 1 1])

julia&gt; fieldnames(Phenomes)
(:entries, :populations, :traits, :phenotypes, :mask)

julia&gt; phenomes.entries = [&quot;entry_1&quot;, &quot;entry_2&quot;];

julia&gt; phenomes.populations = [&quot;pop_A&quot;, &quot;pop_B&quot;];

julia&gt; phenomes.traits = [&quot;height&quot;, &quot;yield&quot;];

julia&gt; phenomes.phenotypes = [200.0 2.5; 150.0 missing];

julia&gt; phenomes.mask = [true true; true false];

julia&gt; phenomes
Phenomes([&quot;entry_1&quot;, &quot;entry_2&quot;], [&quot;pop_A&quot;, &quot;pop_B&quot;], [&quot;height&quot;, &quot;yield&quot;], Union{Missing, Float64}[200.0 2.5; 150.0 missing], Bool[1 1; 1 0])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/all_structs.jl#L59-L97">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.SimulatedEffects" href="#GBCore.SimulatedEffects"><code>GBCore.SimulatedEffects</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>SimulatedEffects struct</strong></p><p>Contains:</p><ol><li>Identification, i.e. trait, year, season, harvest, site, and replication</li><li>Additive environmental effects:<ul><li>year</li><li>season</li><li>site</li></ul></li><li>Environmental interaction effects:<ul><li>seasons<em>x</em>year</li><li>harvests<em>x</em>season<em>x</em>year</li><li>sites<em>x</em>harvest<em>x</em>season<em>x</em>year</li></ul></li><li>Spatial effects including the field layout per year-season-harvest-site combination<ul><li>field_layout</li><li>replications<em>x</em>site<em>x</em>harvest<em>x</em>season<em>x</em>year</li><li>blocks<em>x</em>site<em>x</em>harvest<em>x</em>season<em>x</em>year</li><li>rows<em>x</em>site<em>x</em>harvest<em>x</em>season<em>x</em>year</li><li>cols<em>x</em>site<em>x</em>harvest<em>x</em>season<em>x</em>year</li></ul></li><li>Genetic effects<ul><li>additive_genetic</li><li>dominance_genetic</li><li>epistasis_genetic</li></ul></li><li>GxE effects<ul><li>additive<em>allele</em>x<em>site</em>x<em>harvest</em>x<em>season</em>x_year</li><li>dominance<em>allele</em>x<em>site</em>x<em>harvest</em>x<em>season</em>x_year</li><li>epistasis<em>allele</em>x<em>site</em>x<em>harvest</em>x<em>season</em>x_year</li></ul></li></ol><p><strong>Constructor</strong></p><pre><code class="language-julia hljs">SimulatedEffects()</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/all_structs.jl#L217-L249">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.TEBV" href="#GBCore.TEBV"><code>GBCore.TEBV</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Trial-estimate breeding values (TEBV) struct</strong></p><p>Contains trial-estimated breeding values as generated by <code>analyse(trials::Trials, ...)</code>.</p><p><strong>Fields</strong></p><ul><li><code>traits</code>: names of the traits <code>t</code> traits</li><li><code>formulae</code>: best-fitting formula for each trait</li><li><code>models</code>: best-fitting linear mixed model for each trait</li><li><code>df_BLUEs</code>: vector of data frames of best linear unbiased estimators or fixed effects table of each best fitting model</li><li><code>df_BLUPs</code>: vector of data frames of best linear unbiased predictors or random effects table of each best fitting model</li><li><code>phenomes</code>: vector of Phenomes structs each containing the breeding values</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; tebv = TEBV(traits=[&quot;&quot;], formulae=[&quot;&quot;], models=[MixedModel(@formula(y~1+(1|x)), DataFrame(y=1, x=1))], df_BLUEs=[DataFrame(x=1)], df_BLUPs=[DataFrame(x=1)], phenomes=[Phenomes(n=1, t=1)]);

julia&gt; tebv.traits
1-element Vector{String}:
 &quot;&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/all_structs.jl#L175-L196">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.Trials" href="#GBCore.Trials"><code>GBCore.Trials</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Trials struct</strong></p><p>Contains phenotype data across years, seasons, harvest, sites, populations, replications, blocks, rows, and columns</p><p><strong>Constructor</strong></p><pre><code class="language-julia hljs">Trials(; n::Int64 = 2, p::Int64 = 2)</code></pre><p><strong>Fields</strong></p><ul><li><code>phenotypes</code>: <code>n x t</code> matrix of numeric phenotype data which can have missing values</li><li><code>traits</code>: names of the traits <code>t</code> traits</li><li><code>years</code>: names of the years corresponding to each row in the phenotype matrix</li><li><code>seasons</code>: names of the seasons corresponding to each row in the phenotype matrix</li><li><code>harvests</code>: names of the harvests corresponding to each row in the phenotype matrix</li><li><code>sites</code>: names of the sites corresponding to each row in the phenotype matrix</li><li><code>replications</code>: names of the replications corresponding to each row in the phenotype matrix</li><li><code>blocks</code>: names of the blocks corresponding to each row in the phenotype matrix</li><li><code>rows</code>: names of the rows corresponding to each row in the phenotype matrix</li><li><code>cols</code>: names of the cols corresponding to each row in the phenotype matrix</li><li><code>entries</code>: names of the entries corresponding to each row in the phenotype matrix</li><li><code>populations</code>: names of the populations corresponding to each row in the phenotype matrix</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; trials = Trials(n=1, t=2)
Trials(Union{Missing, Float64}[missing missing], [&quot;&quot;, &quot;&quot;], [&quot;&quot;], [&quot;&quot;], [&quot;&quot;], [&quot;&quot;], [&quot;&quot;], [&quot;&quot;], [&quot;&quot;], [&quot;&quot;], [&quot;&quot;], [&quot;&quot;])

julia&gt; fieldnames(Trials)
(:phenotypes, :traits, :years, :seasons, :harvests, :sites, :replications, :blocks, :rows, :cols, :entries, :populations)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/all_structs.jl#L110-L142">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.:==-Tuple{CV, CV}" href="#Base.:==-Tuple{CV, CV}"><code>Base.:==</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.:(==)(x::CV, y::CV)::Bool</code></pre><p>Equality of CV structs using the hash function defined for CV structs.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; fit = Fit(n=1, l=2);

julia&gt; cv_1 = CV(&quot;replication_1&quot;, &quot;fold_1&quot;, fit, [&quot;population_1&quot;], [&quot;entry_1&quot;], [0.0], [0.0], fit.metrics);

julia&gt; cv_2 = clone(cv_1);

julia&gt; cv_3 = clone(cv_1); cv_3.replication = &quot;other_replication&quot;;

julia&gt; cv_1 == cv_2
true

julia&gt; cv_1 == cv_3
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/cv.jl#L68-L89">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.:==-Tuple{Fit, Fit}" href="#Base.:==-Tuple{Fit, Fit}"><code>Base.:==</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.:(==)(x::Fit, y::Fit)::Bool</code></pre><p>Equality of Fit structs using the hash function defined for Fit structs.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; fit_1 = Fit(n=1, l=4);

julia&gt; fit_2 = Fit(n=1, l=4);

julia&gt; fit_3 = Fit(n=1, l=2);

julia&gt; fit_1 == fit_2
true

julia&gt; fit_1 == fit_3
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/fit.jl#L57-L76">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.:==-Tuple{Genomes, Genomes}" href="#Base.:==-Tuple{Genomes, Genomes}"><code>Base.:==</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.:(==)(x::Genomes, y::Genomes)::Bool</code></pre><p>Equality of Genomes structs using the hash function defined for Genomes structs.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes_1 = genomes = Genomes(n=2,p=4);

julia&gt; genomes_2 = genomes = Genomes(n=2,p=4);

julia&gt; genomes_3 = genomes = Genomes(n=1,p=2);

julia&gt; genomes_1 == genomes_2
true

julia&gt; genomes_1 == genomes_3
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/genomes.jl#L45-L64">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.:==-Tuple{Phenomes, Phenomes}" href="#Base.:==-Tuple{Phenomes, Phenomes}"><code>Base.:==</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.:(==)(x::Phenomes, y::Phenomes)::Bool</code></pre><p>Equality of Phenomes structs using the hash function defined for Phenomes structs.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; phenomes_1 = phenomes = Phenomes(n=2, t=4);

julia&gt; phenomes_2 = phenomes = Phenomes(n=2, t=4);

julia&gt; phenomes_3 = phenomes = Phenomes(n=1, t=2);

julia&gt; phenomes_1 == phenomes_2
true

julia&gt; phenomes_1 == phenomes_3
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/phenomes.jl#L42-L61">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.:==-Tuple{SimulatedEffects, SimulatedEffects}" href="#Base.:==-Tuple{SimulatedEffects, SimulatedEffects}"><code>Base.:==</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.:(==)(x::SimulatedEffects, y::SimulatedEffects)::Bool</code></pre><p>Equality of SimulatedEffects structs using the hash function defined for SimulatedEffects structs.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; effects_1 = SimulatedEffects();

julia&gt; effects_2 = SimulatedEffects();

julia&gt; effects_3 = SimulatedEffects(); effects_3.id[1] = &quot;SOMETHING_ELSE&quot;;

julia&gt; effects_1 == effects_2
true

julia&gt; effects_1 == effects_3
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/simulation/simulate_effects.jl#L76-L95">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.:==-Tuple{TEBV, TEBV}" href="#Base.:==-Tuple{TEBV, TEBV}"><code>Base.:==</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.:(==)(x::TEBV, y::TEBV)::Bool</code></pre><p>Equality of TEBV structs using the hash function defined for TEBV structs.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; tebv_1 = TEBV(traits=[&quot;&quot;], formulae=[&quot;&quot;], models=[MixedModel(@formula(y~1+(1|x)), DataFrame(y=1, x=1))], df_BLUEs=[DataFrame(x=1)], df_BLUPs=[DataFrame(x=1)], phenomes=[Phenomes(n=1,t=1)]);

julia&gt; tebv_2 = TEBV(traits=[&quot;&quot;], formulae=[&quot;&quot;], models=[MixedModel(@formula(y~1+(1|x)), DataFrame(y=1, x=1))], df_BLUEs=[DataFrame(x=1)], df_BLUPs=[DataFrame(x=1)], phenomes=[Phenomes(n=1,t=1)]);

julia&gt; tebv_3 = TEBV(traits=[&quot;SOMETHING_ELSE&quot;], formulae=[&quot;&quot;], models=[MixedModel(@formula(y~1+(1|x)), DataFrame(y=1, x=1))], df_BLUEs=[DataFrame(x=1)], df_BLUPs=[DataFrame(x=1)], phenomes=[Phenomes(n=1,t=1)]);

julia&gt; tebv_1 == tebv_2
true

julia&gt; tebv_1 == tebv_3
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/tebv.jl#L52-L71">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.:==-Tuple{Trials, Trials}" href="#Base.:==-Tuple{Trials, Trials}"><code>Base.:==</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.:(==)(x::Trials, y::Trials)::Bool</code></pre><p>Equality of Trials structs using the hash function defined for Trials structs.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; trials_1 = trials = Trials(n=2, t=4);

julia&gt; trials_2 = trials = Trials(n=2, t=4);

julia&gt; trials_3 = trials = Trials(n=1, t=2);

julia&gt; trials_1 == trials_2
true

julia&gt; trials_1 == trials_3
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/trials.jl#L75-L94">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.filter-Tuple{Genomes, Float64}" href="#Base.filter-Tuple{Genomes, Float64}"><code>Base.filter</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">filter(
    genomes::Genomes;
    maf::Float64,
    max_entry_sparsity::Float64 = 0.0,
    max_locus_sparsity::Float64 = 0.0,
    chr_pos_allele_ids::Union{Missing,Vector{String}} = missing,
)::Genomes</code></pre><p>Filter a Genomes struct by minimum allele frequency, entries sparsity, loci sparsity, and/or vector of locus-allele names</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = simulategenomes(n=100, l=1_000, n_alleles=4, verbose=false);

julia&gt; filtered_genomes_1 = filter(genomes, 0.1);

julia&gt; filtered_genomes_2 = filter(genomes, 0.1, chr_pos_allele_ids=genomes.loci_alleles[1:1000]);

julia&gt; size(genomes.allele_frequencies)
(100, 3000)

julia&gt; size(filtered_genomes_1.allele_frequencies)
(100, 1236)

julia&gt; size(filtered_genomes_2.allele_frequencies)
(100, 388)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/genomes.jl#L617-L645">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.filter-Tuple{Genomes}" href="#Base.filter-Tuple{Genomes}"><code>Base.filter</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">filter(genomes::Genomes)::Genomes</code></pre><p>Filter a Genomes struct using its mask matrix where all rows and columns with at least one false value are excluded</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = simulategenomes(verbose=false); genomes.mask[1:10, 42:100] .= false;
    
julia&gt; filtered_genomes = filter(genomes);

julia&gt; size(filtered_genomes.allele_frequencies)
(90, 9941)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/genomes.jl#L593-L607">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.filter-Tuple{Phenomes}" href="#Base.filter-Tuple{Phenomes}"><code>Base.filter</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">filter(phenomes::Phenomes)::Phenomes</code></pre><p>Filter a Phenomes struct using its mask matrix where all rows and columns with at least one false value are excluded</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; phenomes = Phenomes(n=10, t=3); phenomes.entries = string.(&quot;entry_&quot;, 1:10); phenomes.populations .= &quot;pop_1&quot;; phenomes.traits = [&quot;A&quot;, &quot;B&quot;, &quot;C&quot;]; phenomes.phenotypes = fill(0.0, 10,3);

julia&gt; phenomes.mask .= true; phenomes.mask[6:10, 1] .= false;
    
julia&gt; filtered_phenomes = filter(phenomes);

julia&gt; size(filtered_phenomes.phenotypes)
(5, 2)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/phenomes.jl#L501-L517">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.hash-Tuple{CV, UInt64}" href="#Base.hash-Tuple{CV, UInt64}"><code>Base.hash</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.hash(x::CV, h::UInt)::UInt</code></pre><p>Hash a CV struct using the entries, populations and loci<em>alleles. We deliberately excluded the allele</em>frequencies, and mask for efficiency.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; fit = Fit(n=1, l=2);

julia&gt; cv = CV(&quot;replication_1&quot;, &quot;fold_1&quot;, fit, [&quot;population_1&quot;], [&quot;entry_1&quot;], [0.0], [0.0], fit.metrics);

julia&gt; typeof(hash(cv))
UInt64</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/cv.jl#L29-L44">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.hash-Tuple{Fit, UInt64}" href="#Base.hash-Tuple{Fit, UInt64}"><code>Base.hash</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.hash(x::Fit, h::UInt)::UInt</code></pre><p>Hash a Fit struct using the entries, populations and loci<em>alleles. We deliberately excluded the allele</em>frequencies, and mask for efficiency.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; fit = Fit(n=1, l=2);

julia&gt; typeof(hash(fit))
UInt64</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/fit.jl#L29-L42">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.hash-Tuple{Genomes, UInt64}" href="#Base.hash-Tuple{Genomes, UInt64}"><code>Base.hash</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.hash(x::Genomes, h::UInt)::UInt</code></pre><p>Hash a Genomes struct using the entries, populations and loci<em>alleles. We deliberately excluded the allele</em>frequencies, and mask for efficiency.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = Genomes(n=2, p=2);

julia&gt; typeof(hash(genomes))
UInt64</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/genomes.jl#L25-L38">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.hash-Tuple{Phenomes, UInt64}" href="#Base.hash-Tuple{Phenomes, UInt64}"><code>Base.hash</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.hash(x::Phenomes, h::UInt)::UInt</code></pre><p>Hash a Phenomes struct.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; phenomes = Phenomes(n=2, t=2);

julia&gt; typeof(hash(phenomes))
UInt64</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/phenomes.jl#L24-L36">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.hash-Tuple{SimulatedEffects, UInt64}" href="#Base.hash-Tuple{SimulatedEffects, UInt64}"><code>Base.hash</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.hash(x::SimulatedEffects, h::UInt)::UInt</code></pre><p>Hash a SimulatedEffects struct.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; effects = SimulatedEffects();

julia&gt; typeof(hash(effects))
UInt64</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/simulation/simulate_effects.jl#L1-L13">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.hash-Tuple{TEBV, UInt64}" href="#Base.hash-Tuple{TEBV, UInt64}"><code>Base.hash</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.hash(x::TEBV, h::UInt)::UInt</code></pre><p>Hash a TEBV struct using the traits, formualae and phenomes. We deliberately excluded the models, df<em>BLUEs, and df</em>BLUPs for efficiency.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; tebv = TEBV(traits=[&quot;&quot;], formulae=[&quot;&quot;], models=[MixedModel(@formula(y~1+(1|x)), DataFrame(y=1, x=1))], df_BLUEs=[DataFrame(x=1)], df_BLUPs=[DataFrame(x=1)], phenomes=[Phenomes(n=1,t=1)]);

julia&gt; typeof(hash(tebv))
UInt64</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/tebv.jl#L32-L45">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.hash-Tuple{Trials, UInt64}" href="#Base.hash-Tuple{Trials, UInt64}"><code>Base.hash</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Base.hash(x::Trials, h::UInt)::UInt</code></pre><p>Hash a Trials struct.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; trials = Trials(n=2, t=2);

julia&gt; typeof(hash(trials))
UInt64</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/trials.jl#L33-L45">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.merge-Tuple{Genomes, Genomes}" href="#Base.merge-Tuple{Genomes, Genomes}"><code>Base.merge</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">merge(
    genomes::Genomes,
    other::Genomes;
    conflict_resolution::Tuple{Float64,Float64} = (0.5, 0.5),
    verbose::Bool = true,
)::Genomes</code></pre><p>Merge two Genomes structs using a tuple of conflict resolution weights</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; n = 100; l = 5_000; n_alleles = 2;

julia&gt; all = simulategenomes(n=n, l=l, n_alleles=n_alleles, verbose=false);

julia&gt; genomes = slice(all, idx_entries=collect(1:Int(floor(n*0.75))), idx_loci_alleles=collect(1:Int(floor(l*(n_alleles-1)*0.75))));

julia&gt; other = slice(all, idx_entries=collect(Int(floor(n*0.50)):n), idx_loci_alleles=collect(Int(floor(l*(n_alleles-1)*0.50)):l*(n_alleles-1)));

julia&gt; merged_genomes = merge(genomes, other, conflict_resolution=(0.75, 0.25), verbose=false);

julia&gt; size(merged_genomes.allele_frequencies)
(100, 5000)

julia&gt; sum(ismissing.(merged_genomes.allele_frequencies))
123725</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/genomes.jl#L780-L808">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.merge-Tuple{Genomes, Phenomes}" href="#Base.merge-Tuple{Genomes, Phenomes}"><code>Base.merge</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">merge(genomes::Genomes, phenomes::Phenomes; keep_all::Bool=true)::Tuple{Genomes,Phenomes}</code></pre><p>Merge a Genomes struct with a Phenomes struct using union or intersection</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = simulategenomes(n=10, verbose=false);

julia&gt; trials, effects = simulatetrials(genomes=slice(genomes, idx_entries=collect(1:5), idx_loci_alleles=collect(1:length(genomes.loci_alleles))), f_add_dom_epi=[0.90 0.05 0.05;], n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=2, verbose=false);

julia&gt; phenomes = Phenomes(n=5, t=1);

julia&gt; phenomes.entries = trials.entries[1:5];

julia&gt; phenomes.populations = trials.populations[1:5];

julia&gt; phenomes.traits = trials.traits;

julia&gt; phenomes.phenotypes = trials.phenotypes[1:5, :];

julia&gt; phenomes.mask .= true;

julia&gt; genomes_merged_1, phenomes_merged_1 = merge(genomes, phenomes, keep_all=true);

julia&gt; size(genomes_merged_1.allele_frequencies), size(phenomes_merged_1.phenotypes)
((10, 10000), (10, 1))

julia&gt; genomes_merged_2, phenomes_merged_2 = merge(genomes, phenomes, keep_all=false);

julia&gt; size(genomes_merged_2.allele_frequencies), size(phenomes_merged_2.phenotypes)
((5, 10000), (5, 1))</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/genomes.jl#L933-L966">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.merge-Tuple{Phenomes, Phenomes}" href="#Base.merge-Tuple{Phenomes, Phenomes}"><code>Base.merge</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">merge(
    phenomes::Phenomes,
    other::Phenomes;
    conflict_resolution::Tuple{Float64,Float64} = (0.5, 0.5),
    verbose::Bool = true,
)::Phenomes</code></pre><p>Merge two Phenomes structs using a tuple of conflict resolution weights</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; all = Phenomes(n=10, t=3); all.entries = string.(&quot;entry_&quot;, 1:10); all.traits = [&quot;A&quot;, &quot;B&quot;, &quot;C&quot;]; all.phenotypes = rand(10,3);

julia&gt; phenomes = slice(all, idx_entries=collect(1:7), idx_traits=[1,2]);

julia&gt; other = slice(all, idx_entries=collect(5:10), idx_traits=[2,3]);

julia&gt; merged_phenomes = merge(phenomes, other, conflict_resolution=(0.75, 0.25), verbose=false);

julia&gt; size(merged_phenomes.phenotypes)
(10, 3)

julia&gt; sum(ismissing.(merged_phenomes.phenotypes))
7</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/phenomes.jl#L526-L552">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Base.sum-Tuple{SimulatedEffects}" href="#Base.sum-Tuple{SimulatedEffects}"><code>Base.sum</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">sum(effects::SimulatedEffects)::Tuple{Int64, Int64, Int64}</code></pre><p>Sum up the simulated effects to generate the simulated phenotype values</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; effects = SimulatedEffects();

julia&gt; sum(effects)
1-element Vector{Float64}:
 0.0

julia&gt; effects.additive_genetic[1] = pi;

julia&gt; sum(effects)
1-element Vector{Float64}:
 3.141592653589793</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/simulation/simulate_effects.jl#L137-L156">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.addcompositetrait-Tuple{Phenomes}" href="#GBCore.addcompositetrait-Tuple{Phenomes}"><code>GBCore.addcompositetrait</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">addcompositetrait(phenomes::Phenomes; composite_trait_name::String, formula_string::String)::Phenomes</code></pre><p>Add a composite trait from existing traits in a Phenomes struct</p><pre><code class="language-julia-repl hljs">julia&gt; phenomes = Phenomes(n=10, t=3); phenomes.entries = string.(&quot;entry_&quot;, 1:10); phenomes.populations .= &quot;pop_1&quot;; phenomes.traits = [&quot;A&quot;, &quot;B&quot;, &quot;C&quot;]; phenomes.phenotypes = rand(10,3);

julia&gt; phenomes_new = addcompositetrait(phenomes, composite_trait_name = &quot;some_wild_composite_trait&quot;, formula_string = &quot;A&quot;);

julia&gt; phenomes_new.phenotypes[:, end] == phenomes.phenotypes[:, 1]
true

julia&gt; phenomes_new = addcompositetrait(phenomes, composite_trait_name = &quot;some_wild_composite_trait&quot;, formula_string = &quot;(A^B) + (C/A) - sqrt(abs(B-A)) + log(1.00 + C)&quot;);

julia&gt; phenomes_new.phenotypes[:, end] == (phenomes.phenotypes[:,1].^phenomes.phenotypes[:,2]) .+ (phenomes.phenotypes[:,3]./phenomes.phenotypes[:,1]) .- sqrt.(abs.(phenomes.phenotypes[:,2].-phenomes.phenotypes[:,1])) .+ log.(1.00 .+ phenomes.phenotypes[:,3])
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/phenomes.jl#L720-L738">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.addcompositetrait-Tuple{Trials}" href="#GBCore.addcompositetrait-Tuple{Trials}"><code>GBCore.addcompositetrait</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">addcompositetrait(trials::Trials; composite_trait_name::String, formula_string::String)::Trials</code></pre><p>Add a composite trait from existing traits in a Trials struct</p><pre><code class="language-julia-repl hljs">julia&gt; trials, _ = simulatetrials(genomes = simulategenomes(verbose=false), verbose=false);

julia&gt; trials_new = addcompositetrait(trials, composite_trait_name = &quot;some_wild_composite_trait&quot;, formula_string = &quot;trait_1&quot;);

julia&gt; trials_new.phenotypes[:, end] == trials.phenotypes[:, 1]
true

julia&gt; trials_new = addcompositetrait(trials, composite_trait_name = &quot;some_wild_composite_trait&quot;, formula_string = &quot;(trait_1^(trait_2/100)) + (trait_3/trait_1) - sqrt(abs(trait_2-trait_1)) + log(1.00 + trait_3)&quot;);

julia&gt; trials_new.phenotypes[:, end] == (trials.phenotypes[:,1].^(trials.phenotypes[:,2]/100)) .+ (trials.phenotypes[:,3]./trials.phenotypes[:,1]) .- sqrt.(abs.(trials.phenotypes[:,2].-trials.phenotypes[:,1])) .+ log.(1.00 .+ trials.phenotypes[:,3])
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/trials.jl#L309-L327">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.analyse" href="#GBCore.analyse"><code>GBCore.analyse</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">analyse(
    trials::Trials,
    formula_string::String = &quot;&quot;;
    traits::Union{Nothing,Vector{String}} = nothing,
    max_levels::Int64 = 100,
    max_time_per_model::Int64 = 60,
    covariates_continuous::Union{Nothing,Vector{String}} = nothing,
    verbose::Bool = true,
)::TEBV</code></pre><p><strong>Analyse trials</strong></p><p><strong>Arguments</strong></p><ul><li><code>trials</code>: Trials struct </li><li><code>max_levels</code>: maximum number of non-entry factor levels to include in the linear mixed models (default = 100)</li><li><code>max_time_per_model</code>: maximum time in seconds for fitting each linear mixed model (default = 60)</li><li><code>verbose</code>: Show trials analysis progress bar? (default = true)</li></ul><p><strong>Outputs</strong></p><ul><li><code>TEBV</code> struct containing the trait names, the best fitting formulae, models, BLUEs, and BLUPs for each trait</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; trials, _simulated_effects = simulatetrials(genomes = simulategenomes(n=10, verbose=false), n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=10, verbose=false);

julia&gt; tebv_1 = analyse(trials, &quot;trait_1 ~ 1 + (1|entries)&quot;, max_levels=50, verbose=false);

julia&gt; tebv_1.traits
3-element Vector{String}:
 &quot;trait_1&quot;
 &quot;trait_2&quot;
 &quot;trait_3&quot;

julia&gt; tebv_2 = analyse(trials, max_levels=50, verbose=false);

julia&gt; mean(tebv_2.phenomes[1].phenotypes) &lt; mean(tebv_2.phenomes[2].phenotypes)
true

julia&gt; trials = addcompositetrait(trials, composite_trait_name = &quot;covariate&quot;, formula_string = &quot;(trait_1 + trait_2) / (trait_3 + 0.0001)&quot;);

julia&gt; tebv_3 = Suppressor.@suppress analyse(trials, &quot;y ~ 1 + covariate + entries + (1|blocks)&quot;, max_levels=50, verbose=false);

julia&gt; mean(tebv_3.phenomes[1].phenotypes) &lt; mean(tebv_3.phenomes[2].phenotypes)
true

julia&gt; tebv_4 = Suppressor.@suppress analyse(trials, max_levels=50, covariates_continuous=[&quot;covariate&quot;], verbose=false);

julia&gt; mean(tebv_4.phenomes[1].phenotypes) &lt; mean(tebv_4.phenomes[2].phenotypes)
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/tebv.jl#L691-L742">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.analyse-Tuple{DataFrames.DataFrame}" href="#GBCore.analyse-Tuple{DataFrames.DataFrame}"><code>GBCore.analyse</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">analyse(
    df::DataFrame; 
    formulae::Vector{String},
    idx_parallel_models::Vector{Int64},
    idx_iterative_models::Vector{Int64},
    max_time_per_model::Int64 = 60,
    verbose::Bool=false)::Tuple{String, Any, DataFrame, DataFrame, Phenomes}</code></pre><p>Fit univarite (one trait) linear mixed models to extract the effects of the entries the best fitting model.</p><p>We have the following guiding principles:</p><ul><li>Avoid over-parameterisation we&#39;ll have enough of that with the genomic prediction models</li><li>We will fit mixed models with unstructure variance-covariance matrix of the random effects</li><li>We prefer REML over ML</li><li>We compare BLUEs vs BLUPs of entries</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; trials, _ = simulatetrials(genomes = simulategenomes(verbose=false), verbose=false);

julia&gt; df = tabularise(trials);

julia&gt; formulae, n_levels = trialsmodelsfomulae!(df; trait = &quot;trait_1&quot;, max_levels = 10);

julia&gt; idx_parallel_models::Vector{Int64} = findall(n_levels .&lt;= (15));

julia&gt; idx_iterative_models::Vector{Int64} = findall((n_levels .&lt;= (15)) .!= true);

julia&gt; formula_string, model, df_BLUEs, df_BLUPs, phenomes = analyse(df, formulae=formulae, idx_parallel_models=idx_parallel_models, idx_iterative_models=idx_iterative_models);

julia&gt; length(phenomes.entries) == length(unique(df.entries))
true

julia&gt; df_2 = df[(df.years .== df.years[1]) .&amp;&amp; (df.harvests .== df.harvests[1]) .&amp;&amp; (df.seasons .== df.seasons[1]) .&amp;&amp; (df.sites .== df.sites[1]) .&amp;&amp; (df.replications .== df.replications[1]), :];

julia&gt; formula_string_2, model_2, df_BLUEs_2, df_BLUPs_2, phenomes_2 = analyse(df_2, formulae=[&quot;trait_1 ~ 1 + 1|entries&quot;]);

julia&gt; cor(phenomes_2.phenotypes[sortperm(phenomes_2.entries),1], df_2.trait_1[sortperm(df_2.entries)]) == 1.00
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/tebv.jl#L414-L456">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.checkdims-Tuple{CV}" href="#GBCore.checkdims-Tuple{CV}"><code>GBCore.checkdims</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">checkdims(cv::CV)::Bool</code></pre><p>Check dimension compatibility of the fields of the CV struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; fit = Fit(n=1, l=2);

julia&gt; cv = CV(&quot;replication_1&quot;, &quot;fold_1&quot;, fit, [&quot;population_1&quot;], [&quot;entry_1&quot;], [0.0], [0.0], fit.metrics);

julia&gt; checkdims(cv)
true

julia&gt; cv.validation_y_true = [0.0, 0.0];

julia&gt; checkdims(cv)
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/cv.jl#L95-L114">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.checkdims-Tuple{Fit}" href="#GBCore.checkdims-Tuple{Fit}"><code>GBCore.checkdims</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">checkdims(fit::Fit)::Bool</code></pre><p>Check dimension compatibility of the fields of the Fit struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; fit = Fit(n=1, l=4);

julia&gt; checkdims(fit)
true

julia&gt; fit.b_hat_labels = [&quot;chr1\t1\tA|T\tA&quot;];

julia&gt; checkdims(fit)
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/fit.jl#L82-L99">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.checkdims-Tuple{Genomes}" href="#GBCore.checkdims-Tuple{Genomes}"><code>GBCore.checkdims</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">checkdims(genomes::Genomes)::Bool</code></pre><p>Check dimension compatibility of the fields of the Genomes struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = Genomes(n=2,p=4);

julia&gt; checkdims(genomes)
false

julia&gt; genomes.entries = [&quot;entry_1&quot;, &quot;entry_2&quot;];

julia&gt; genomes.loci_alleles = [&quot;chr1\t1\tA|T\tA&quot;, &quot;chr1\t2\tC|G\tG&quot;, &quot;chr2\t3\tA|T\tA&quot;, &quot;chr2\t4\tG|T\tG&quot;];

julia&gt; checkdims(genomes)
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/genomes.jl#L70-L89">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.checkdims-Tuple{Phenomes}" href="#GBCore.checkdims-Tuple{Phenomes}"><code>GBCore.checkdims</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">checkdims(y::Phenomes)::Bool</code></pre><p>Check dimension compatibility of the fields of the Phenomes struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; y = Phenomes(n=2, t=2);

julia&gt; checkdims(y)
false

julia&gt; y.entries = [&quot;entry_1&quot;, &quot;entry_2&quot;];

julia&gt; y.traits = [&quot;trait_1&quot;, &quot;trait_2&quot;];

julia&gt; checkdims(y)
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/phenomes.jl#L67-L86">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.checkdims-Tuple{SimulatedEffects}" href="#GBCore.checkdims-Tuple{SimulatedEffects}"><code>GBCore.checkdims</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">checkdims(effects::SimulatedEffects)::Bool</code></pre><p>Check dimension compatibility of the fields of the SimulatedEffects struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; effects = SimulatedEffects();

julia&gt; checkdims(effects)
true

julia&gt; effects.id = [&quot;beaking_change&quot;];

julia&gt; checkdims(effects)
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/simulation/simulate_effects.jl#L101-L118">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.checkdims-Tuple{TEBV}" href="#GBCore.checkdims-Tuple{TEBV}"><code>GBCore.checkdims</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">checkdims(y::TEBV)::Bool</code></pre><p>Check dimension compatibility of the fields of the TEBV struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; tebv = TEBV(traits=[&quot;&quot;], formulae=[&quot;&quot;], models=[MixedModel(@formula(y~1+(1|x)), DataFrame(y=1, x=1))], df_BLUEs=[DataFrame(x=1)], df_BLUPs=[DataFrame(x=1)], phenomes=[Phenomes(n=1,t=1)]);

julia&gt; checkdims(tebv)
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/tebv.jl#L77-L89">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.checkdims-Tuple{Trials}" href="#GBCore.checkdims-Tuple{Trials}"><code>GBCore.checkdims</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">checkdims(trials::Trials)::Bool</code></pre><p>Check dimension compatibility of the fields of the Trials struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; trials = Trials(n=1, t=2);

julia&gt; trials.entries = [&quot;entry_1&quot;]; trials.traits = [&quot;trait_1&quot;, &quot;trait_2&quot;];

julia&gt; checkdims(trials)
true

julia&gt; trials.entries = [&quot;entering_2_entries&quot;, &quot;instead_of_just_1&quot;];

julia&gt; checkdims(trials)
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/trials.jl#L100-L119">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.clone-Tuple{CV}" href="#GBCore.clone-Tuple{CV}"><code>GBCore.clone</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">clone(x::CV)::CV</code></pre><p>Clone a CV object</p><p><strong>Example</strong></p><pre><code class="language-julia-repl hljs">julia&gt; fit = Fit(n=1, l=2);

julia&gt; cv = CV(&quot;replication_1&quot;, &quot;fold_1&quot;, fit, [&quot;population_1&quot;], [&quot;entry_1&quot;], [0.0], [0.0], fit.metrics);

julia&gt; copy_cv = clone(cv)
CV(&quot;replication_1&quot;, &quot;fold_1&quot;, Fit(&quot;&quot;, [&quot;&quot;, &quot;&quot;], [0.0, 0.0], &quot;&quot;, [&quot;&quot;], [&quot;&quot;], [0.0], [0.0], Dict(&quot;&quot; =&gt; 0.0)), [&quot;population_1&quot;], [&quot;entry_1&quot;], [0.0], [0.0], Dict(&quot;&quot; =&gt; 0.0))</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/cv.jl#L1-L15">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.clone-Tuple{Fit}" href="#GBCore.clone-Tuple{Fit}"><code>GBCore.clone</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">clone(x::Fit)::Fit</code></pre><p>Clone a Fit object</p><p><strong>Example</strong></p><pre><code class="language-julia-repl hljs">julia&gt; fit = Fit(n=1, l=2);

julia&gt; copy_fit = clone(fit)
Fit(&quot;&quot;, [&quot;&quot;, &quot;&quot;], [0.0, 0.0], &quot;&quot;, [&quot;&quot;], [&quot;&quot;], [0.0], [0.0], Dict(&quot;&quot; =&gt; 0.0))</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/fit.jl#L1-L13">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.clone-Tuple{Genomes}" href="#GBCore.clone-Tuple{Genomes}"><code>GBCore.clone</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">clone(x::Genomes)::Genomes</code></pre><p>Clone a Genomes object</p><p><strong>Example</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = Genomes(n=2, p=2);

julia&gt; copy_genomes = clone(genomes)
Genomes([&quot;&quot;, &quot;&quot;], [&quot;&quot;, &quot;&quot;], [&quot;&quot;, &quot;&quot;], Union{Missing, Float64}[missing missing; missing missing], Bool[1 1; 1 1])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/genomes.jl#L1-L13">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.clone-Tuple{Phenomes}" href="#GBCore.clone-Tuple{Phenomes}"><code>GBCore.clone</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">clone(x::Phenomes)::Phenomes</code></pre><p>Clone a Phenomes object</p><p><strong>Example</strong></p><pre><code class="language-julia-repl hljs">julia&gt; phenomes = Phenomes(n=2, t=2);

julia&gt; copy_phenomes = clone(phenomes)
Phenomes([&quot;&quot;, &quot;&quot;], [&quot;&quot;, &quot;&quot;], [&quot;&quot;, &quot;&quot;], Union{Missing, Float64}[missing missing; missing missing], Bool[1 1; 1 1])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/phenomes.jl#L1-L13">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.clone-Tuple{TEBV}" href="#GBCore.clone-Tuple{TEBV}"><code>GBCore.clone</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">clone(x::TEBV)::TEBV</code></pre><p>Clone a TEBV object</p><p><strong>Example</strong></p><pre><code class="language-julia-repl hljs">julia&gt; tebv = TEBV(traits=[&quot;&quot;], formulae=[&quot;&quot;], models=[MixedModel(@formula(y~1+(1|x)), DataFrame(y=1, x=1))], df_BLUEs=[DataFrame(x=1)], df_BLUPs=[DataFrame(x=1)], phenomes=[Phenomes(n=1,t=1)]);

julia&gt; copy_tebv = clone(tebv);

julia&gt; copy_tebv.traits == tebv.traits
true

julia&gt; copy_tebv.phenomes == tebv.phenomes
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/tebv.jl#L1-L18">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.clone-Tuple{Trials}" href="#GBCore.clone-Tuple{Trials}"><code>GBCore.clone</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">clone(x::Trials)::Trials</code></pre><p>Clone a Trials object</p><p><strong>Example</strong></p><pre><code class="language-julia-repl hljs">julia&gt; trials = Trials(n=2, t=2);

julia&gt; copy_trials = clone(trials)
Trials(Union{Missing, Float64}[missing missing; missing missing], [&quot;&quot;, &quot;&quot;], [&quot;&quot;, &quot;&quot;], [&quot;&quot;, &quot;&quot;], [&quot;&quot;, &quot;&quot;], [&quot;&quot;, &quot;&quot;], [&quot;&quot;, &quot;&quot;], [&quot;&quot;, &quot;&quot;], [&quot;&quot;, &quot;&quot;], [&quot;&quot;, &quot;&quot;], [&quot;&quot;, &quot;&quot;], [&quot;&quot;, &quot;&quot;])</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/trials.jl#L1-L13">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.countlevels-Tuple{DataFrames.DataFrame}" href="#GBCore.countlevels-Tuple{DataFrames.DataFrame}"><code>GBCore.countlevels</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">countlevels(df::DataFrame; column_names::Vector{String})::Int64</code></pre><p>Count the total number of factor levels across the column names provided.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/tebv.jl#L159-L163">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.dimensions-Tuple{Genomes}" href="#GBCore.dimensions-Tuple{Genomes}"><code>GBCore.dimensions</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">dimensions(genomes::Genomes)::Dict{String, Int64}</code></pre><p>Count the number of entries, populations, loci-alleles combination, loci, maximum number of alleles per locus, and number of missing data in the Genomes struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = simulategenomes(n=100, l=1_000, n_alleles=4, verbose=false);

julia&gt; dimensions(genomes)
Dict{String, Int64} with 7 entries:
  &quot;n_entries&quot;      =&gt; 100
  &quot;n_chr&quot;          =&gt; 7
  &quot;n_loci&quot;         =&gt; 1000
  &quot;n_loci_alleles&quot; =&gt; 3000
  &quot;n_populations&quot;  =&gt; 1
  &quot;n_missing&quot;      =&gt; 0
  &quot;max_n_alleles&quot;  =&gt; 4</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/genomes.jl#L103-L122">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.dimensions-Tuple{Phenomes}" href="#GBCore.dimensions-Tuple{Phenomes}"><code>GBCore.dimensions</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">dimensions(phenomes::Phenomes)::Dict{String, Int64}</code></pre><p>Count the number of entries, populations, and traits in the Phenomes struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; phenomes = Phenomes(n=10, t=3); phenomes.entries = string.(&quot;entry_&quot;, 1:10); phenomes.populations .= &quot;pop_1&quot;; phenomes.traits = [&quot;A&quot;, &quot;B&quot;, &quot;C&quot;]; phenomes.phenotypes = fill(0.0, 10,3);

julia&gt; dimensions(phenomes)
Dict{String, Int64} with 8 entries:
  &quot;n_total&quot;       =&gt; 30
  &quot;n_zeroes&quot;      =&gt; 30
  &quot;n_nan&quot;         =&gt; 0
  &quot;n_entries&quot;     =&gt; 10
  &quot;n_traits&quot;      =&gt; 3
  &quot;n_inf&quot;         =&gt; 0
  &quot;n_populations&quot; =&gt; 1
  &quot;n_missing&quot;     =&gt; 0</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/phenomes.jl#L100-L120">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.dimensions-Tuple{TEBV}" href="#GBCore.dimensions-Tuple{TEBV}"><code>GBCore.dimensions</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">dimensions(tebv::TEBV)::Dict{String, Int64}</code></pre><p>Count the number of entries, populations, and traits in the TEBV struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; tebv = TEBV(traits=[&quot;trait_1&quot;], formulae=[&quot;trait_1 ~ 1 + 1|entries&quot;], models=[MixedModel(@formula(y~1+(1|x)), DataFrame(y=1, x=1))], df_BLUEs=[DataFrame(x=1)], df_BLUPs=[DataFrame(x=1)], phenomes=[Phenomes(n=1,t=1)]);

julia&gt; dimensions(tebv)
Dict{String, Int64} with 8 entries:
  &quot;n_total&quot;       =&gt; 1
  &quot;n_zeroes&quot;      =&gt; 0
  &quot;n_nan&quot;         =&gt; 0
  &quot;n_entries&quot;     =&gt; 1
  &quot;n_traits&quot;      =&gt; 1
  &quot;n_inf&quot;         =&gt; 0
  &quot;n_populations&quot; =&gt; 1
  &quot;n_missing&quot;     =&gt; 1</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/tebv.jl#L102-L122">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.dimensions-Tuple{Trials}" href="#GBCore.dimensions-Tuple{Trials}"><code>GBCore.dimensions</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">dimensions(trials::Trials)::Dict{String, Int64}</code></pre><p>Count the number of entries, populations, and traits in the Trials struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; trials = Trials(n=1, t=2);

julia&gt; trials.entries = [&quot;entry_1&quot;]; trials.traits = [&quot;trait_1&quot;, &quot;trait_2&quot;];

julia&gt; dimensions(trials)
Dict{String, Int64} with 16 entries:
  &quot;n_zeroes&quot;       =&gt; 0
  &quot;n_harvests&quot;     =&gt; 1
  &quot;n_nan&quot;          =&gt; 0
  &quot;n_entries&quot;      =&gt; 1
  &quot;n_traits&quot;       =&gt; 2
  &quot;n_seasons&quot;      =&gt; 1
  &quot;n_rows&quot;         =&gt; 1
  &quot;n_blocks&quot;       =&gt; 1
  &quot;n_missing&quot;      =&gt; 2
  &quot;n_inf&quot;          =&gt; 0
  &quot;n_total&quot;        =&gt; 2
  &quot;n_replications&quot; =&gt; 1
  &quot;n_years&quot;        =&gt; 1
  &quot;n_sites&quot;        =&gt; 1
  &quot;n_cols&quot;         =&gt; 1
  &quot;n_populations&quot;  =&gt; 1</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/trials.jl#L139-L169">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.distances-Tuple{Genomes}" href="#GBCore.distances-Tuple{Genomes}"><code>GBCore.distances</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">distances(
    genomes::Genomes; 
    distance_metrics::Vector{String}=[&quot;euclidean&quot;, &quot;correlation&quot;, &quot;mad&quot;, &quot;rmsd&quot;, &quot;χ²&quot;]
    idx_loci_alleles::Union{Nothing, Vector{Int64}} = nothing,
)::Tuple{Vector{String}, Vector{String}, Dict{String, Matrix{Float64}}}</code></pre><p>Estimate pairwise distances between loci_alleles and entries.  Sparsity leading to less than 2 pairs will yield -Inf values in the resulting matrices.  Matrices with how many pairs were used to estimate the distance and correlation matrices are also included as well as the locus-allele and entry names.</p><pre><code class="language-julia-repl hljs">julia&gt; genomes = simulategenomes(n=100, l=1_000, n_alleles=4, verbose=false);

julia&gt; (loci_alleles, entries, dist) = distances(genomes, distance_metrics=[&quot;correlation&quot;, &quot;χ²&quot;]);

julia&gt; sort(string.(keys(dist))) == [&quot;entries|correlation&quot;, &quot;entries|counts&quot;, &quot;entries|χ²&quot;, &quot;loci_alleles|correlation&quot;, &quot;loci_alleles|counts&quot;, &quot;loci_alleles|χ²&quot;]
true

julia&gt; C = dist[&quot;entries|correlation&quot;]; C[diagind(C)] == repeat([1], length(genomes.entries))
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/genomes.jl#L248-L270">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.distances-Tuple{Phenomes}" href="#GBCore.distances-Tuple{Phenomes}"><code>GBCore.distances</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">distances(
    phenomes::Phenomes; 
    distance_metrics::Vector{String}=[&quot;euclidean&quot;, &quot;correlation&quot;, &quot;mad&quot;, &quot;rmsd&quot;, &quot;χ²&quot;],
    standardise_traits::Bool = false,
)::Tuple{String, Matrix{Float64}}Tuple{Vector{String}, Vector{String}, Dict{String, Matrix{Float64}}}</code></pre><p>Estimate pairwise distances between traits and entries.  Sparsity leading to less than 2 pairs will yield -Inf values in the resulting matrices.  Matrices with how many pairs were used to estimate the distance and correlation matrices are also included as well as the trait and entry names. Note that prior to computing correlations between entries, the phenotype matrix was standardised, i.e. each trait was transformed to have a mean of 0 and standard deviation of 1.</p><pre><code class="language-julia-repl hljs">julia&gt; phenomes = Phenomes(n=10, t=3); phenomes.entries = string.(&quot;entry_&quot;, 1:10); phenomes.populations .= &quot;pop_1&quot;; phenomes.traits = [&quot;A&quot;, &quot;B&quot;, &quot;C&quot;]; phenomes.phenotypes = rand(10,3); phenomes.phenotypes[2,2] = missing;

julia&gt; (traits, entries, dist) = distances(phenomes, distance_metrics=[&quot;correlation&quot;, &quot;χ²&quot;]);

julia&gt; sort(string.(keys(dist))) == [&quot;entries|correlation&quot;, &quot;entries|counts&quot;, &quot;entries|χ²&quot;, &quot;traits|correlation&quot;, &quot;traits|counts&quot;, &quot;traits|χ²&quot;]
true

julia&gt; C = dist[&quot;entries|correlation&quot;]; C[diagind(C)] == repeat([1], length(phenomes.entries))
true

julia&gt; dist[&quot;traits|counts&quot;][:, 2] == dist[&quot;traits|counts&quot;][2, :] == repeat([9], length(phenomes.traits))
true

julia&gt; dist[&quot;entries|counts&quot;][:, 2] == dist[&quot;entries|counts&quot;][2, :] == repeat([2], length(phenomes.entries))
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/phenomes.jl#L138-L167">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.extractphenomes-Tuple{TEBV}" href="#GBCore.extractphenomes-Tuple{TEBV}"><code>GBCore.extractphenomes</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">extractphenomes(tebv::TEBV)::Phenomes</code></pre><p>Extract Phenomes from TEBV</p><pre><code class="language-julia-repl hljs">julia&gt; trials, _simulated_effects = simulatetrials(genomes = simulategenomes(n=10, verbose=false), n_years=1, n_seasons=1, n_harvests=1, n_sites=1, n_replications=10, verbose=false);

julia&gt; tebv = analyse(trials, max_levels=50, verbose=false);

julia&gt; phenomes = extractphenomes(tebv);

julia&gt; phenomes.traits == [&quot;trait_1&quot;, &quot;trait_2&quot;, &quot;trait_3&quot;]
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/tebv.jl#L887-L902">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.extractphenomes-Tuple{Trials}" href="#GBCore.extractphenomes-Tuple{Trials}"><code>GBCore.extractphenomes</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">extractphenomes(trials::Trials)::Phenomes</code></pre><p>Extract Phenomes from Trials struct. Each trait-by-environment variables combination make up the traits in the resulting Phenomes struct. The trait names start with trait name in Trials suffixed by the trait-by-environment variables combinations. If there is a single environment variables combination, then no suffix is added to the trait name.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; trials, _ = simulatetrials(genomes = simulategenomes(verbose=false), verbose=false);

julia&gt; phenomes = extractphenomes(trials);

julia&gt; size(phenomes.phenotypes)
(100, 384)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/trials.jl#L235-L252">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.loci-Tuple{Genomes}" href="#GBCore.loci-Tuple{Genomes}"><code>GBCore.loci</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">loci(genomes::Genomes)::Tuple{Vector{String},Vector{Int64},Vector{Int64},Vector{Int64}}</code></pre><p>Extract chromosome names, positions, start and end indexes of each locus across loci</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = simulategenomes(n=100, l=1_000, n_alleles=4, verbose=false);

julia&gt; chromosomes, positions, loci_ini_idx, loci_fin_idx = loci(genomes);

julia&gt; length(chromosomes), length(positions), length(loci_ini_idx), length(loci_fin_idx)
(1000, 1000, 1000, 1000)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/genomes.jl#L200-L214">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.loci_alleles-Tuple{Genomes}" href="#GBCore.loci_alleles-Tuple{Genomes}"><code>GBCore.loci_alleles</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">loci_alleles(genomes::Genomes)::Tuple{Vector{String},Vector{Int64},Vector{String}}</code></pre><p>Extract chromosomes, positions, and alleles across loci-allele combinations</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = simulategenomes(n=100, l=1_000, n_alleles=4, verbose=false);

julia&gt; chromosomes, positions, alleles = loci_alleles(genomes);

julia&gt; length(chromosomes), length(positions), length(alleles)
(3000, 3000, 3000)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/genomes.jl#L168-L182">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.plot" href="#GBCore.plot"><code>GBCore.plot</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">plot(genomes::Genomes)::Nothing</code></pre><p>Plot allele frequencies</p><p><strong>Examples</strong></p><pre><code class="nohighlight hljs">julia&gt; genomes = simulategenomes(n=100, l=1_000, n_alleles=4, verbose=false);

julia&gt; GBCore.plot(genomes)
</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/genomes.jl#L436-L448">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.plot" href="#GBCore.plot"><code>GBCore.plot</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">plot(fit::Fit, distribution::Any=[TDist(1), Normal()][1], α::Float64=0.05)</code></pre><p>Manhattan plot</p><p><strong>Examples</strong></p><pre><code class="nohighlight hljs">julia&gt; distribution = [TDist(1), Normal()][2];

julia&gt; fit = Fit(n=100, l=10_000); fit.b_hat = rand(distribution, 10_000);  α = 0.05;

julia&gt; GBCore.plot(fit);
</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/fit.jl#L112-L126">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.plot-Tuple{Phenomes}" href="#GBCore.plot-Tuple{Phenomes}"><code>GBCore.plot</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">plot(phenomes::Phenomes)::Nothing</code></pre><p>Plot histogram/s of the trait value/s and a heatmap of trait correlations</p><p><strong>Examples</strong></p><pre><code class="nohighlight hljs">julia&gt; phenomes = Phenomes(n=10, t=3); phenomes.entries = string.(&quot;entry_&quot;, 1:10); phenomes.populations .= &quot;pop_1&quot;; phenomes.traits = [&quot;A&quot;, &quot;B&quot;, &quot;C&quot;]; phenomes.phenotypes = rand(10,3);

julia&gt; GBCore.plot(phenomes);
</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/phenomes.jl#L338-L350">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.plot-Tuple{TEBV}" href="#GBCore.plot-Tuple{TEBV}"><code>GBCore.plot</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">plot(tebv::TEBV)</code></pre><p>Plot TEBV output by plotting the resulting Phenomes struct</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/tebv.jl#L953-L957">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.plot-Tuple{Trials}" href="#GBCore.plot-Tuple{Trials}"><code>GBCore.plot</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">plot(trials::Trials)::Nothing</code></pre><p>Plot histogram/s of the trait value/s and a heatmap of trait correlations across the entire trial. Additionally, plot mean trait values per year, season, harvest, site, replications, row, column, and population for each trait.</p><p><strong>Examples</strong></p><pre><code class="nohighlight hljs">julia&gt; trials, _ = simulatetrials(genomes = simulategenomes(verbose=false), verbose=false);

julia&gt; GBCore.plot(trials);
</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/trials.jl#L372-L385">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.simulateeffects-Tuple{}" href="#GBCore.simulateeffects-Tuple{}"><code>GBCore.simulateeffects</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Simulate effects</strong></p><p>Sample <code>p</code> x <code>q</code> effects from a multivariate normal distribution with <code>μ~Exp(λ)</code> and <code>Σ=μμ&#39;</code></p><p><strong>Arguments</strong></p><ul><li><code>p</code>: number of correlated effects to simulate (default = 2)</li><li><code>q</code>: number times to simulate the correlated effects from the same distribution (default = 1)</li><li><code>λ</code>: parameter of the exponential distritbution from which the means will be sampled from (default = 1.00)</li><li><code>seed</code>: randomisation seed (default = 42)</li></ul><p><strong>Output</strong></p><ul><li><code>p</code> x <code>q</code> matrix of correlated effects</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; θ::Matrix{Float64} = simulateeffects();

julia&gt; sum(abs.(θ - [-0.0886501800782904; -0.596478483888422])) &lt; 0.00001
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/simulation/simulate_effects.jl#L168-L190">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.simulategenomes-Tuple{}" href="#GBCore.simulategenomes-Tuple{}"><code>GBCore.simulategenomes</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Simulate genomes</strong></p><p><strong>Arguments</strong></p><ul><li><code>n</code>: number of entries (default = 100)</li><li><code>l</code>: number of loci (default = 10_000)</li><li><code>n_chroms</code>: number of chromosomes (default = 7)</li><li><code>n_alleles</code>: number of alleles per locus (default = 2)</li><li><code>max_pos</code>: total length of the genome in base-pairs (bp) (default = 135<em>000</em>000)</li><li><code>ld_corr_50perc_kb</code>: distance in bp at which linkage expressed as correlation between a pair of loci is at 50% (default = 1_000)</li><li><code>μ_β_params</code>: the shape parameters of the Beta distribution from which the mean allele frequencies will be sampled  (default = (0.5, 0.5); U-shaped distribution; you may use (2.0, 2.0) for a bell-shaped distribution)</li><li><code>sparsity</code>: Proportion of missing data (default = 0.0)</li><li><code>seed</code>: psuedo-random number generator seed for replicability (default = 42)</li><li><code>verbose</code>: Show progress bar and plot the linkage heatmap into an svg file? (default = true)</li></ul><p><strong>Output</strong></p><ul><li><code>Genomes</code></li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = simulategenomes(n=100, l=10_000, n_alleles=3, verbose=false);

julia&gt; length(genomes.entries)
100

julia&gt; length(genomes.populations)
100

julia&gt; length(genomes.loci_alleles)
20000

julia&gt; size(genomes.allele_frequencies)
(100, 20000)

julia&gt; mean(ismissing.(genomes.allele_frequencies))
0.0

julia&gt; rng::TaskLocalRNG = Random.seed!(123);

julia&gt; idx = StatsBase.sample(rng, range(1, 20_000, step=2), 250, replace = false, ordered = true);

julia&gt; correlations = StatsBase.cor(genomes.allele_frequencies[:, idx]);

julia&gt; correlations[10,10] == 1.00
true

julia&gt; correlations[10,10] &gt; correlations[10,250]
true

julia&gt; genomes = simulategenomes(n=100, l=10_000, n_alleles=3, sparsity=0.25, verbose=false);

julia&gt; mean(ismissing.(genomes.allele_frequencies))
0.25</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/simulation/simulate_genomes.jl#L1-L56">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.simulategenomiceffects-Tuple{}" href="#GBCore.simulategenomiceffects-Tuple{}"><code>GBCore.simulategenomiceffects</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Simulate genomic effects</strong></p><p>Simulate additive, dominance, and epistatic effects</p><p><strong>Arguments</strong></p><ul><li><code>genomes</code>: Genome struct includes the <code>n</code> entries x <code>p</code> loci-alleles combinations (<code>p</code> = <code>l</code> loci x <code>a-1</code> alleles)</li><li><code>f_additive</code>: proportion of the <code>l</code> loci with non-zero additive effects on the phenotype</li><li><code>f_dominance</code>: proportion of the <code>l*f_additive</code> additive effects loci with additional dominance effects</li><li><code>f_epistasis</code>: proportion of the <code>l*f_additive</code> additive effects loci with additional epistasis effects</li></ul><p><strong>Outputs</strong></p><ul><li><code>n</code> x <code>3</code> matrix of additive, dominance and epistasis effects per entry</li><li><code>p</code> x <code>3</code> matrix of additive, dominance and epistasis effects per locus-allele combination</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes::Genomes = simulategenomes(n=100, l=2_000, n_alleles=3, verbose=false);

julia&gt; G, B = simulategenomiceffects(genomes=genomes, f_additive=0.05, f_dominance=0.75, f_epistasis=0.25);

julia&gt; size.([G, B])
2-element Vector{Tuple{Int64, Int64}}:
 (100, 3)
 (4000, 3)

julia&gt; sum(B .!= 0.0, dims=1)
1×3 Matrix{Int64}:
 200  75  50</code></pre><p><strong>Details</strong></p><p>The additive, dominance, and epistasis allele effects share a common exponential distribution (<code>λ=1</code>) from which  the mean of the effects (<code>μ</code>) are sampled, and the covariance matrix is derived (<code>Σ = μ * μ&#39;</code>;  where if <code>det(Σ)≈0</code> then we iteratively add 1.00 to the diagonals until it becomes invertible or 10 iterations  finishes and throws an error). The non-additive or epistasis allele effects were simulated by multiplying the allele  frequencies of all possible unique pairs of epistasis alleles and their effects.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/simulation/simulate_effects.jl#L213-L250">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.simulatetrials-Tuple{}" href="#GBCore.simulatetrials-Tuple{}"><code>GBCore.simulatetrials</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p><strong>Simulate trials</strong></p><p><strong>Arguments</strong></p><ul><li><code>genomes</code>: Genome struct includes the <code>n</code> entries x <code>p</code> loci-alleles combinations (<code>p</code> = <code>l</code> loci x <code>a-1</code> alleles)</li><li><code>f_add_dom_epi</code>: <code>n_traits</code> x 3 numeric matrix of loci proportion with additive, dominance and epistasis effects, i.e. each column refers to:<ul><li><code>f_additive</code>: proportion of the <code>l</code> loci with non-zero additive effects on the phenotype</li><li><code>f_dominance</code>: proportion of the <code>l*f_additive</code> additive effects loci with additional dominance effects</li><li><code>f_epistasis</code>: proportion of the <code>l*f_additive</code> additive effects loci with additional epistasis effects</li><li>(default = [0.01 0.25 0.10; 0.05 0.50 0.25; 0.10 0.25 0.00])</li></ul></li><li><code>n_years</code>: Number of years (default = 2)</li><li><code>n_seasons</code>: Number of seasons (default = 4)</li><li><code>n_harvests</code>: Number of harvests (default = 2)</li><li><code>n_sites</code>: Number of sites (default = 4)</li><li><code>n_replications</code>: Number of replications (default = 2)</li><li><code>n_blocks</code>: Number of blocks across the entire field layout (default = missing)</li><li><code>n_rows</code>: Number of rows across the entire field layout (default = missing)</li><li><code>n_cols</code>: Number of columns across the entire field layout (default = missing)</li><li><code>proportion_of_variance</code>: <code>9</code> x <code>n_traits</code> numeric matrix of scaled/non-scaled proportion of variances allocated to   genetic and environmental effects (default = missing; values will be sampled from a uniform distribution  followed by a biased sample on the first row, i.e. additive effects row).  The rows correspond to the variance allocated to:<ol><li>additive genetic effects</li><li>dominance genetic effects</li><li>epistasis genetic effects</li><li>years effects</li><li>seasons effects</li><li>sites effects</li><li>environmental interactions</li><li>spatial interactions</li><li>GxE interactiions</li></ol></li><li><code>seed</code>: Randomisation seed (default = 42)</li><li><code>sparsity</code>: Proportion of missing data (default = 0.0)</li><li><code>verbose</code>: Show trials simulation progress bar? (default = true)</li></ul><p><strong>Outputs</strong></p><ul><li><code>Trials</code> struct of simulated phenotype data</li><li>Vector of <code>SimulatedEffects</code> each corresponding to each trait-year-season-harvest-site-replication combination</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes::Genomes = simulategenomes(n=100, l=2_000, n_alleles=3, verbose=false);

julia&gt; trials::Trials, vector_of_effects::Array{SimulatedEffects,1} = simulatetrials(genomes=genomes, sparsity=0.25, verbose=false);

julia&gt; size(trials.phenotypes)
(12800, 3)

julia&gt; size(trials.traits)
(3,)

julia&gt; unique(trials.entries) == genomes.entries
true

julia&gt; unique(trials.populations) == unique(genomes.populations)
true

julia&gt; abs(mean(ismissing.(trials.phenotypes)) - 0.25) &lt; 0.00001
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/simulation/simulate_trials.jl#L1-L63">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.slice-Tuple{Genomes}" href="#GBCore.slice-Tuple{Genomes}"><code>GBCore.slice</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">slice(
    genomes::Genomes; 
    idx_entries::Union{Nothing, Vector{Int64}} = nothing,
    idx_loci_alleles::Union{Nothing, Vector{Int64}} = nothing,
)::Genomes</code></pre><p>Slice a Genomes struct by specifing indixes of entries and loci-allele combinations</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; genomes = simulategenomes(n=100, l=1_000, n_alleles=4, verbose=false);

julia&gt; sliced_genomes = slice(genomes, idx_entries=collect(1:10); idx_loci_alleles=collect(1:300));

julia&gt; dimensions(sliced_genomes)
Dict{String, Int64} with 7 entries:
  &quot;n_entries&quot;      =&gt; 10
  &quot;n_chr&quot;          =&gt; 1
  &quot;n_loci&quot;         =&gt; 100
  &quot;n_loci_alleles&quot; =&gt; 300
  &quot;n_populations&quot;  =&gt; 1
  &quot;n_missing&quot;      =&gt; 0
  &quot;max_n_alleles&quot;  =&gt; 4</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/genomes.jl#L517-L542">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.slice-Tuple{Phenomes}" href="#GBCore.slice-Tuple{Phenomes}"><code>GBCore.slice</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">slice(
    phenomes::Phenomes;
    idx_entries::Union{Nothing, Vector{Int64}} = nothing,
    idx_traits::Union{Nothing, Vector{Int64}} = nothing,
)::Phenomes</code></pre><p>Slice a Phenomes struct by specifing indexes of entries and traits</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; phenomes = Phenomes(n=10, t=3); phenomes.entries = string.(&quot;entry_&quot;, 1:10); phenomes.populations .= &quot;pop_1&quot;; phenomes.traits = [&quot;A&quot;, &quot;B&quot;, &quot;C&quot;]; phenomes.phenotypes = fill(0.0, 10,3);

julia&gt; sliced_phenomes = slice(phenomes, idx_entries=collect(1:5); idx_traits=collect(2:3));

julia&gt; dimensions(sliced_phenomes)
Dict{String, Int64} with 8 entries:
  &quot;n_total&quot;       =&gt; 10
  &quot;n_zeroes&quot;      =&gt; 10
  &quot;n_nan&quot;         =&gt; 0
  &quot;n_entries&quot;     =&gt; 5
  &quot;n_traits&quot;      =&gt; 2
  &quot;n_inf&quot;         =&gt; 0
  &quot;n_populations&quot; =&gt; 1
  &quot;n_missing&quot;     =&gt; 0</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/phenomes.jl#L423-L449">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.slice-Tuple{Trials}" href="#GBCore.slice-Tuple{Trials}"><code>GBCore.slice</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">slice(
    trials::Trials; 
    traits::Union{Nothing, Vector{String}} = nothing,
    populations::Union{Nothing, Vector{String}} = nothing,
    entries::Union{Nothing, Vector{String}} = nothing,
    years::Union{Nothing, Vector{String}} = nothing,
    harvests::Union{Nothing, Vector{String}} = nothing,
    seasons::Union{Nothing, Vector{String}} = nothing,
    sites::Union{Nothing, Vector{String}} = nothing,
    replications::Union{Nothing, Vector{String}} = nothing,
    blocks::Union{Nothing, Vector{String}} = nothing,
    rows::Union{Nothing, Vector{String}} = nothing,
    cols::Union{Nothing, Vector{String}} = nothing,
)::Trials</code></pre><p>Slice a Trials struct by specifing indexes of entries and traits</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; trials, _ = simulatetrials(genomes = simulategenomes(verbose=false), verbose=false);

julia&gt; sliced_trials = slice(trials, traits=trials.traits[2:3], years=[unique(trials.years)[1]], seasons=unique(trials.seasons)[2:3]);

julia&gt; dimensions(sliced_trials)
Dict{String, Int64} with 16 entries:
  &quot;n_zeroes&quot;       =&gt; 0
  &quot;n_harvests&quot;     =&gt; 2
  &quot;n_nan&quot;          =&gt; 0
  &quot;n_entries&quot;      =&gt; 100
  &quot;n_traits&quot;       =&gt; 2
  &quot;n_seasons&quot;      =&gt; 2
  &quot;n_blocks&quot;       =&gt; 10
  &quot;n_rows&quot;         =&gt; 10
  &quot;n_missing&quot;      =&gt; 0
  &quot;n_inf&quot;          =&gt; 0
  &quot;n_total&quot;        =&gt; 6400
  &quot;n_replications&quot; =&gt; 2
  &quot;n_years&quot;        =&gt; 1
  &quot;n_sites&quot;        =&gt; 4
  &quot;n_cols&quot;         =&gt; 20
  &quot;n_populations&quot;  =&gt; 1</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/trials.jl#L475-L518">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.summarise-Tuple{Vector{CV}}" href="#GBCore.summarise-Tuple{Vector{CV}}"><code>GBCore.summarise</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">summarise(cvs::Vector{CV})::Tuple{DataFrame,DataFrame}</code></pre><p>Summarise a vector of CV structs into:</p><ul><li>a data frame of mean metrics, and</li><li>a data frame of mean and standard deviation of phenotype predictions per entry, trait and model.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; fit_1 = Fit(n=1, l=2); fit_1.metrics = Dict(&quot;cor&quot; =&gt; 0.0, &quot;rmse&quot; =&gt; 1.0); fit_1.trait = &quot;trait_1&quot;;

julia&gt; cv_1 = CV(&quot;replication_1&quot;, &quot;fold_1&quot;, fit_1, [&quot;population_1&quot;], [&quot;entry_1&quot;], [0.0], [0.0], fit_1.metrics);

julia&gt; fit_2 = Fit(n=1, l=2); fit_2.metrics = Dict(&quot;cor&quot; =&gt; 1.0, &quot;rmse&quot; =&gt; 0.0); fit_2.trait = &quot;trait_1&quot;;

julia&gt; cv_2 = CV(&quot;replication_2&quot;, &quot;fold_2&quot;, fit_2, [&quot;population_2&quot;], [&quot;entry_2&quot;], [0.0], [0.0], fit_2.metrics);

julia&gt; cvs = [cv_1, cv_2];

julia&gt; df_summary, df_summary_per_entry = summarise(cvs);

julia&gt; size(df_summary)
(2, 6)

julia&gt; size(df_summary_per_entry)
(2, 8)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/cv.jl#L267-L295">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.tabularise" href="#GBCore.tabularise"><code>GBCore.tabularise</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">tabularise(trials::Trials)::DataFrame</code></pre><p>Export the Trials structs into a DataFrames.DataFrame struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; fit = Fit(n=100, l=10_000); fit.b_hat = rand(10_000); fit.model=&quot;some_model&quot;; fit.trait=&quot;some_trait&quot;; 

julia&gt; fit.metrics = Dict(&quot;cor&quot; =&gt; rand(), &quot;rmse&quot; =&gt; rand()); fit.populations .= &quot;pop_1&quot;;

julia&gt; df = tabularise(fit);

julia&gt; size(df)
(10000, 6)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/fit.jl#L147-L163">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.tabularise-Tuple{Phenomes}" href="#GBCore.tabularise-Tuple{Phenomes}"><code>GBCore.tabularise</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">tabularise(phenomes::Phenomes)::DataFrame</code></pre><p>Export the Phenomes structs into a DataFrames.DataFrame struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; phenomes = Phenomes(n=10, t=3); phenomes.entries = string.(&quot;entry_&quot;, 1:10); phenomes.populations .= &quot;pop_1&quot;; phenomes.traits = [&quot;A&quot;, &quot;B&quot;, &quot;C&quot;]; phenomes.phenotypes = fill(0.0, 10,3);

julia&gt; tabularise(phenomes)
10×6 DataFrame
 Row │ id     entries   populations  A         B         C        
     │ Int64  String    String       Float64?  Float64?  Float64? 
─────┼────────────────────────────────────────────────────────────
   1 │     1  entry_1   pop_1             0.0       0.0       0.0
   2 │     2  entry_2   pop_1             0.0       0.0       0.0
   3 │     3  entry_3   pop_1             0.0       0.0       0.0
   4 │     4  entry_4   pop_1             0.0       0.0       0.0
   5 │     5  entry_5   pop_1             0.0       0.0       0.0
   6 │     6  entry_6   pop_1             0.0       0.0       0.0
   7 │     7  entry_7   pop_1             0.0       0.0       0.0
   8 │     8  entry_8   pop_1             0.0       0.0       0.0
   9 │     9  entry_9   pop_1             0.0       0.0       0.0
  10 │    10  entry_10  pop_1             0.0       0.0       0.0</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/phenomes.jl#L671-L696">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.tabularise-Tuple{Trials}" href="#GBCore.tabularise-Tuple{Trials}"><code>GBCore.tabularise</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">tabularise(trials::Trials)::DataFrame</code></pre><p>Export the Trials structs into a DataFrames.DataFrame struct</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; trials, _ = simulatetrials(genomes = simulategenomes(verbose=false), verbose=false);

julia&gt; df = tabularise(trials);

julia&gt; size(df)
(12800, 14)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/trials.jl#L195-L209">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.tabularise-Tuple{Vector{CV}}" href="#GBCore.tabularise-Tuple{Vector{CV}}"><code>GBCore.tabularise</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">tabularise(cvs::Vector{CV})::Tuple{DataFrame,DataFrame}</code></pre><p>Export a vector of CV structs into data frames of metrics across entries and per validation entry</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; fit_1 = Fit(n=1, l=2); fit_1.metrics = Dict(&quot;cor&quot; =&gt; 0.0, &quot;rmse&quot; =&gt; 1.0); fit_1.trait = &quot;trait_1&quot;;

julia&gt; cv_1 = CV(&quot;replication_1&quot;, &quot;fold_1&quot;, fit_1, [&quot;population_1&quot;], [&quot;entry_1&quot;], [0.0], [0.0], fit_1.metrics);

julia&gt; fit_2 = Fit(n=1, l=2); fit_2.metrics = Dict(&quot;cor&quot; =&gt; 1.0, &quot;rmse&quot; =&gt; 0.0); fit_2.trait = &quot;trait_1&quot;;

julia&gt; cv_2 = CV(&quot;replication_2&quot;, &quot;fold_2&quot;, fit_2, [&quot;population_2&quot;], [&quot;entry_2&quot;], [0.0], [0.0], fit_2.metrics);

julia&gt; cvs = [cv_1, cv_2];

julia&gt; df_across_entries, df_per_entry = tabularise(cvs);

julia&gt; names(df_across_entries)
8-element Vector{String}:
 &quot;training_population&quot;
 &quot;validation_population&quot;
 &quot;trait&quot;
 &quot;model&quot;
 &quot;replication&quot;
 &quot;fold&quot;
 &quot;cor&quot;
 &quot;rmse&quot;

julia&gt; df_across_entries[!, [:cor, :rmse]]
2×2 DataFrame
 Row │ cor      rmse    
     │ Float64  Float64 
─────┼──────────────────
   1 │     0.0      1.0
   2 │     1.0      0.0

julia&gt; names(df_per_entry)
9-element Vector{String}:
 &quot;training_population&quot;
 &quot;validation_population&quot;
 &quot;entry&quot;
 &quot;trait&quot;
 &quot;model&quot;
 &quot;replication&quot;
 &quot;fold&quot;
 &quot;y_true&quot;
 &quot;y_pred&quot;

julia&gt; df_per_entry[!, [:entry, :y_true, :y_pred]]
2×3 DataFrame
 Row │ entry    y_true   y_pred  
     │ String   Float64  Float64 
─────┼───────────────────────────
   1 │ entry_1      0.0      0.0
   2 │ entry_2      0.0      0.0</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/cv.jl#L128-L186">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.trialsmodelsfomulae!-Tuple{DataFrames.DataFrame}" href="#GBCore.trialsmodelsfomulae!-Tuple{DataFrames.DataFrame}"><code>GBCore.trialsmodelsfomulae!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">trialsmodelsfomulae!(df::DataFrame; trait::String, max_levels::Int64 = 100)::Vector{String}</code></pre><p>Define formulae for the mixed models to fit on the tabularised <code>Trials</code> struct.     - appends interaction effects intto <code>df</code>     - returns:         + a vector of formulae as strings         + a vector of the total number of non-entry factor levels</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; trials, _simulated_effects = simulatetrials(genomes = simulategenomes(verbose=false), verbose=false);

julia&gt; df = tabularise(trials);

julia&gt; size(df)
(12800, 14)

julia&gt; formulae, n_levels = trialsmodelsfomulae!(df, trait=&quot;trait_1&quot;);

julia&gt; size(df)
(12800, 134)

julia&gt; length(formulae)
76

julia&gt; sum(n_levels .== sort(n_levels))
76</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/tebv.jl#L188-L217">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.@string2formula-Tuple{Any}" href="#GBCore.@string2formula-Tuple{Any}"><code>GBCore.@string2formula</code></a> — <span class="docstring-category">Macro</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">string2formula(x)</code></pre><p>Macro to <code>Meta.parse</code> a string into a formula. TODO: Refactor to make more idiomatic.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/tebv.jl#L177-L182">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GBCore.@stringevaluation-Tuple{Any}" href="#GBCore.@stringevaluation-Tuple{Any}"><code>GBCore.@stringevaluation</code></a> — <span class="docstring-category">Macro</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">stringevaluation(x)</code></pre><p>Macro to <code>Meta.parse</code> a string of formula.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/GenomicBreeding/GBCore.jl/blob/67e82a084065054d30597288744a28b5128666b9/src/phenomes.jl#L711-L715">source</a></section></article></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.1 on <span class="colophon-date" title="Monday 17 February 2025 03:12">Monday 17 February 2025</span>. Using Julia version 1.11.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
